### DISCUSSION TITLE
Free Form Deformation DAFOAM

### DISCUSSION THREAD

**Discussion by Janfoam** [mdolab/dafoam#85] [Tutorials]
Hi, 

I am looking into how to generate the FFD points with DAFoam and tried it before, but don't fully understand the procedure. I know from the literature that is available on DAFoam, the FFD points should completely cover the geometry. Some explanations are also given in: https://dafoam.github.io/mydoc_get_started_runscript.html My questions are: 

- How do you select the points? From the example taken from the tutorial, the points appear to all start from zero. This is not very clear to me. Shouldn't this be along the domain outside of the airfoil?
-  How do you know how many FFD points to choose?
- How do you know whether to work with one of two domains with the FFD points? 
- What checks are in place to ensure there are no kinks or overlapping?
- The steps I have been doing so far based on the 

![image](https://user-images.githubusercontent.com/76228791/105076576-839d2500-5a8b-11eb-8ac4-36dc3ed1fe08.png)

![image](https://user-images.githubusercontent.com/76228791/105076685-a92a2e80-5a8b-11eb-8087-d0cce3a4a5a8.png)

**Comment by friedenhe** [comment]
- The first entry is the FFD block number. In this case, you have only one block, that is why it is always zero (check the DPW4 case where two blocks are used; one for the wing and one for the tail), the 2nd entry is the 8 corner indices, and the 3rd entry is the x,y,z coordinates of the corners.
- You have to do trial-and-error. We usually use 20 points for each wing section.
- Not quite sure what you meant, especially the two domains part
- Load the FFD points in Paraview. In version 2.2.1, we added a new utility to convert the plot3D file (*.xyz) format to the Tecplot format, such that you can load the FFD point in Paraview more easily (Paraview sometimes crashes when loading Plot3D files). To use it, just copy over the dafoam/dafoam/scripts/dafoam_plot3d2tecplot.py file and run `python dafoam_plot3d2tecplot.py wingFFD.xyz wingFFDTecplot.dat`

BTW: Are you interested in putting together a guide for the FFD generation and add it to the DAFoam website? This will be very helpful for others.

**Reply by Janfoam** [reply]
Hi friedenhe, Thanks for your responses. 

A few more questions on this: 
- When you say trial and error with the FFD points - what criteria is being used to guide the trial and error efforts? For example, if one decides to use 40 points instead of 20. I already read the article behind the FFD development: G. K. W. Kenway, Kennedy, G. J., and Martins, J. R. R. A., “A CAD-Free Approach to High-Fidelity Aerostructural Optimization”, but did not pick up any info about how to perform this trial and error.
- What I meant with one or two domains was whether to use one block or two blocks. For example in the DPW4 case, there are two blocks. How does one know that one block/two blocks should be used on the same geometry.
-  Thanks for the suggestion regarding the DPW4 tutorial. The setup for the FFD points is a lot different than in the case with one block. I'll study it a bit more to see how the formulas for the coordinates on the left are derived and get back with questions. 
- _Load the FFD points in Paraview. In version 2.2.1, we added a new utility to convert the plot3D file (*.xyz) format to the Tecplot format, such that you can load the FFD point in Paraview more easily (Paraview sometimes crashes when loading Plot3D files). To use it, just copy over the dafoam/dafoam/scripts/dafoam_plot3d2tecplot.py file and run python dafoam_plot3d2tecplot.py wingFFD.xyz wingFFDTecplot.dat_ - ok I'll get back on this once I'm able try the simulation. After the last run with DAFoam the administrator had to recompile so will check the status and see.

Thanks. Sure, I am interested in putting a guide together.  I am not an FFD expert, but will see the best and most accurate way to present the information once doubts are clarified. The suggested points to consider would be:
             - Theoretical aspects so that there is a strong foundation to understand the theory and approach to FFD points and how this is 
                featured in DAFoam. 
             - Practical aspects applied to specific geometries. In my case I would present a wind turbine blade. 
             - Tips and tricks to get a better feel of good practices 
             - Extra bibliography from reading
    
Please confirm if the suggested approach is fine and your thoughts on alternatives.
 
![image](https://user-images.githubusercontent.com/76228791/105422800-a7579b00-5c44-11eb-9514-6f1d7366949f.png)

**Reply by friedenhe** [reply]
- Trial and error means running optimization with different numbers of FFD points and compare the optimized shapes and drag reductions among them
- If your design surfaces are separated, like the wing and tail, you can use two separate FFD boxes (blocks) to cover them. If your design surface is just one patch, no need to use multi-block FFDs

You suggested documentation sounds good to me!

**Reply by TonkomoLLC** [reply]
Hello,
I am also learning how to generate the FFD points with DAFoam. Is it possible to post the genFFD.py file for the U-bend tutorial in the [tutorials GitHub site](https://github.com/DAFoam/tutorials)?
Thanks!

**Reply by friedenhe** [reply]
The FFD for the U bend channel was generated in ICEM-CFD, so, unfortunately, we don't have the genFFD.py file for this case.

**Reply by TonkomoLLC** [reply]
Understood - Thanks @friedenhe for the quick reply.

**Comment by Janfoam** [comment]
Hello, 

We have been using the mesh from the DAFoam's incompressible case. After running the simulation we noticed that the optimized geometry is outside of the FFD control points( please see iteration 50) Is this correct? Usually the optimized geometry would stay within the FFD control points. Right now it would appear as if the FFD control points remain fixed the entire time, but this is not the case- they also change hence why the FFD control box should change. Is there a way to see what the new FFD control points are after the optimization?

- What is the criteria used to generate the additional FFD points?

- Are the FFD normal to the geometry and what is the consequence of having this restriction? We know that with FFD, only small changes in the geometry is permissible - is this the only consequence of having the restriction regarding the normal.

- We will be using a different mesh from the one that is in the case. Do we just need to use the blockMesh command to have DAFoam read our mesh? 

![image](https://user-images.githubusercontent.com/76228791/113771764-190f9300-9724-11eb-82e2-530e1bad7e1a.png)
![image](https://user-images.githubusercontent.com/76228791/113771809-29c00900-9724-11eb-9efb-9f9a65620f2e.png)

**Reply by friedenhe** [reply]
During the optimization, the FFD points will be changed by the optimizer. By default, the deformed FFD will not be output. But if you git clone the latest version of the DAFoam repo, you can set `writeDeformedFFDs: True` in the runScript.py (daOptions), the deformed FFDs will be written to the disk during optimization (see this commit https://github.com/mdolab/dafoam/commit/24cce9c737c29fa8b0cb4c1e5c05d98a27accfb2). I don't follow your questions in the first two bullets (we don't generate new FFD points, I am not sure what do you mean by saying normal to the geometry). If you want to use a new mesh, you need to first make sure the native simpleFoam solver runs, then the mesh should be ready to use in DAFoam.

**Reply by Janfoam** [reply]
Hi friedenhe, 

Some efforts were made by the administrator to install the version of DAFoam that makes it possible to print out the FFD points. However, I am getting the message below. Can you please explain what could be the issue here? It appears the correct DAFoam version is still not installed or am I missing something?




![image](https://user-images.githubusercontent.com/76228791/121739513-0470c200-cafc-11eb-88cd-9cd27839562a.png)

**Reply by friedenhe** [reply]
That is right, you need v2.2.5.

**Reply by Janfoam** [reply]
Hi friedenhe, 

The administrator was able to install the latest version now. However, I am getting the message below. Can you please say what the cause could be? We have updated the solvers, done the necessary troubleshooting, but the error persists. Many thanks

"/home/user/dafoam/packages/miniconda3/lib/python3.6/site-packages/dafoam-2.2.5-py3.6.egg/dafoam/pyDAFoam.py", line 2162, in _initSolver
ModuleNotFoundError: No module named 'dafoam.pyDASolverIncompressible'

**Reply by friedenhe** [reply]
It seems that the installation is not done properly, e.g., maybe you didn't do `pip install .` For airfoil cases, you can run it on your laptop/desktop using the Docker image (we support Windows, Linux, and MacOS). There is no need to use an HPC.

**Comment by hosseinnegahban** [comment]
Hi,
I have imported my mesh with good quality from ICEM and the mesh generation was OK in DAFoam. But in running optimization, I receive this error. what is the nature of this error? 
My second question is that how could we create FFD in ICEM and export it to DAFoam?
![Untitled](https://user-images.githubusercontent.com/97816713/149645652-1b45f0db-80e4-4335-bb8e-230ee30f18b7.png)

**Reply by friedenhe** [reply]
You need to adjust the leList and teList, they were probably too close and no longer completely within the airfoil. So that DAFoam returns an error for the volume and thickness constraints.

**Reply by hosseinnegahban** [reply]
Thank you very much for your answer. It solved the problem. However, I have encountered another error as follows (I have increased the primalMinResTol but not worked, and I can not find primalMinResTolDiff  in runScript to increase

![Untitled](https://user-images.githubusercontent.com/97816713/149819090-25ec59a5-e4a2-4538-8074-85f24a87d931.png)

**Reply by hosseinnegahban** [reply]
The problem was related to mesh. It is solved now.

**Reply by hosseinnegahban** [reply]
Regarding Optimization algorithms, are we free to use any optimization algorithms like particle swarm, genetic algorithm and so on?
If yes, how should we configure them with the default code?

**Comment by sachinsreedhar** [comment]
1.How to export mesh file in xyz format from icem cfd?
2. How to create .xyz file if I know the point co ordinates

**Reply by sreekarrs** [reply]
I think this was answered here #260

**Reply by sachinsreedhar** [reply]
Is there any way to generate .xyz file from a cad tool?

Secondly I can get the point coordinates for the FFD points. How to make .xyz file from this?

**Reply by friedenhe** [reply]
1. xyz are mesh files, so I don't think CAD can generate it.
2. yes. please take a look at an existing xyz file, it is in structure mesh format.

**Reply by sachinsreedhar** [reply]
Thank you for the reply.
Could that be created using any opensource software?

**Reply by friedenhe** [reply]
no.. but you can use our genFFD.py in the FFD folder to generate simple FFD boxes.

**Reply by sachinsreedhar** [reply]
Thanks .but how to create surface for constraints. Similar to JBC Hull tutorial. File name hullcurve.xyz.

**Reply by friedenhe** [reply]
Yeah, for complex geometries, you have to use ICEM. We are not aware of a good open-source tool that can generate plot3D files easily.

**Reply by sachinsreedhar** [reply]
What are the steps to generate plot3d in ICEM ?

**Comment by sachinsreedhar** [comment]
Is there any tutorial on how to generate hullcurve.xyz file?

**Comment by friedenhe** [comment]
Please check ICEM's tutorial on how to generate a body-fitted structured mesh and output it as plot3d.
